accepted/2024/trusted-publishers-oidc-for-nuget-push.technical.md

```diff
@@ -0,0 +1,253 @@
+# Trusted Publishers, using OpenID Connect for NuGet push (technical)
+
+- Author: Joel Verhagen ([@joelverhagen](https://github.com/joelverhagen) on GitHub)
+- Issue: [NuGet/NuGetGallery#9332](https://github.com/NuGet/NuGetGallery/issues/9332)
+
+This is the technical description of the experience described in [Trusted Publishers, using OpenID Connect for NuGet
+push](trusted-publishers-oidc-for-nuget-push.md). The content is split to clearly separate the functional/UX description
+(the other doc) from the technical description (this doc). This document acts as a supporting document to the other so
+that other one should be read first.
+
+Much of the technical explanation is described in [Trusted Publishers for All Package
+Repositories](https://repos.openssf.org/trusted-publishers-for-all-package-repositories). I will expand on certain
+technical details which are particularly interesting or specific to NuGet.
+
+## Validation of the GitHub OIDC token
+
+The most critical step in this design is understanding how NuGet.org will validate that an incoming OIDC token is
+acceptable to trade for a short-lived NuGet API key.
+
+The following checks are made:
+
+- The token is issued by a known Trusted Publisher (only GitHub Actions at this time)
+- The token is validated per all JWT rules, such as but not limited to:
+  - valid signature via JWKS
+  - validate duration (`nbf` and `exp` claims)
+  - only used once, via `jti`
+  - valid `aud` claim, being `nuget`
```

Comment:
Note that although it's not true for GHA, with other IdPs it's possible to generate/provide an OIDC token with multiple `aud` claims which is supported by the OIDC spec: https://openid.net/specs/openid-connect-core-1_0.html#:~:text=REQUIRED.%20Audience(s,case%2Dsensitive%20string.

You may want to additionally verify that _only_ the `nuget` audience is included in the token, to ensure that users aren't potentially sharing nuget-enabled OIDC tokens with additional audiences.

Reply:
Ah, nice call out. I didn't know multiple audiences was possible. Yes, there should be exactly one and it should be a well-defined "nuget" value.